packages/wdio-browserstack-service/src/service.ts

```diff
@@ -0,0 +1,240 @@
+     * will fail
+     */
+    beforeSession (config: WebdriverIO.Config) {
+        if (!config.user) {
+            config.user = 'NotSetUser'
+        }
+
+        if (!config.key) {
+            config.key = 'NotSetKey'
+        }
+        this._config.user = config.user
+        this._config.key = config.key
+    }
+
+    before(caps: Capabilities, specs: string[], browser: Browser) {
+        this._browser = browser
+
+        // Ensure capabilities are not null in case of multiremote
+        const capabilities = this._browser.capabilities || {}
+        if (capabilities.app || this._caps.app) {
+            this._sessionBaseUrl = 'https://api-cloud.browserstack.com/app-automate/sessions'
+        }
+
+        this._scenariosThatRan = []
+
+        return this._printSessionURL()
+    }
+
+    beforeSuite (suite: WebdriverIO.Suite) {
+        this._fullTitle = suite.title
+        return this._updateJob({ name: this._fullTitle })
+    }
+
+    beforeFeature(uri: string, feature: Feature) {
+        this._fullTitle = feature.document.feature.name
+        return this._updateJob({ name: this._fullTitle })
+    }
+
+    afterTest(test: WebdriverIO.Test, context: Context, results: WebdriverIO.TestResult) {
+        const { error, passed } = results
+
+        this._fullTitle = (
+            /**
+             * Jasmine
+             */
+            test.fullName ||
+            /**
+             * Mocha
+             */
+            `${test.parent} - ${test.title}`
+        )
+
+        if (!passed) {
+            this._failReasons.push((error && error.message) || 'Unknown Error')
+        }
+    }
+
+    after(result: number) {
+        // For Cucumber: Checks scenarios that ran (i.e. not skipped) on the session
+        // Only 1 Scenario ran and option enabled => Redefine session name to Scenario's name
+        if (this._preferScenarioName && this._scenariosThatRan.length === 1){
+            this._fullTitle = this._scenariosThatRan.pop()
+        }
+
+        const hasReasons = Boolean(this._failReasons.filter(Boolean).length)
+
+        return this._updateJob({
+            status: result === 0 ? 'passed' : 'failed',
+            name: this._fullTitle,
+            reason: hasReasons ? this._failReasons.join('\n') : undefined
+        })
+    }
+
+    /**
+     * For CucumberJS
+     */
+
+    afterScenario(uri: string, feature: Feature, pickle: Pickle, results: WebdriverIO.TestResult) {
+        let { exception, status } = results
+
+        if (status !== 'skipped') {
+            this._scenariosThatRan.push(pickle.name)
+        }
+
+        if (this._failureStatuses.includes(status)) {
+            exception = exception || (status === 'pending'
+                ? `Some steps/hooks are pending for scenario "${pickle.name}"`
+                : 'Unknown Error')
+
+            this._failReasons.push(exception)
+        }
+    }
+
+    async onReload(oldSessionId: string, newSessionId: string) {
+        const { _browser } = this
```

Comment:
I would prefer to just use `this._browser`

Reply:
👍